tests/app/main/helpers/test_frameworks.py

```diff
@@ -0,0 +1,182 @@
+    content = declaration_content.get_builder()
+    submission = FULL_G7_SUBMISSION.copy()
+    del submission['SQ1-1e']
+
+    assert_equal(get_all_errors(content, submission), {})
+
+
+def test_trading_status_details_error_depends_on_trading_status():
+    content = declaration_content.get_builder()
+    submission = FULL_G7_SUBMISSION.copy()
+    del submission['SQ1-1cii']
+
+    submission['SQ1-1ci'] = "something"
+    assert_equal(get_all_errors(content, submission), {})
+
+    submission['SQ1-1ci'] = "other (please specify)"
+    assert_equal(get_all_errors(content, submission), {'SQ1-1cii': 'answer_required'})
+
+
+def test_trade_registers_details_error_depends_on_trade_registers():
+    content = declaration_content.get_builder()
+    submission = FULL_G7_SUBMISSION.copy()
+    del submission['SQ1-1i-ii']
+
+    submission['SQ1-1i-i'] = False
+    assert_equal(get_all_errors(content, submission), {})
+
+    submission['SQ1-1i-i'] = True
+    assert_equal(get_all_errors(content, submission), {'SQ1-1i-ii': 'answer_required'})
+
+
+def test_licenced_details_error_depends_on_licenced():
+    content = declaration_content.get_builder()
+    submission = FULL_G7_SUBMISSION.copy()
+    del submission['SQ1-1j-ii']
+
+    submission['SQ1-1j-i'] = []
+    assert_equal(get_all_errors(content, submission), {})
+
+    submission['SQ1-1j-i'] = ["something"]
+    assert_equal(get_all_errors(content, submission), {'SQ1-1j-ii': 'answer_required'})
+
+
+def test_no_error_if_no_tax_issues_and_no_details():
+    content = declaration_content.get_builder()
+    submission = FULL_G7_SUBMISSION.copy()
+
+    submission['SQ4-1a'] = False
+    submission['SQ4-1b'] = False
+    del submission['SQ4-1c']
+
+    assert_equal(get_all_errors(content, submission), {})
+
+
+def test_error_if_tax_issues_and_no_details():
+    content = declaration_content.get_builder()
+    submission = FULL_G7_SUBMISSION.copy()
+
+    del submission['SQ4-1c']
+
+    submission['SQ4-1a'] = True
+    submission['SQ4-1b'] = False
+    assert_equal(get_all_errors(content, submission), {'SQ4-1c': 'answer_required'})
+
+    submission['SQ4-1a'] = False
+    submission['SQ4-1b'] = True
+    assert_equal(get_all_errors(content, submission), {'SQ4-1c': 'answer_required'})
+
+
+def test_invalid_duns_number_causes_error():
+    content = declaration_content.get_builder()
+    submission = FULL_G7_SUBMISSION.copy()
+
+    submission['SQ1-1g'] = 'invalid'
+    assert_equal(get_all_errors(content, submission), {'SQ1-1g': 'invalid_format'})
+
+
+def test_invalid_vat_number_causes_error():
+    content = declaration_content.get_builder()
+    submission = FULL_G7_SUBMISSION.copy()
+
+    submission['SQ1-1g'] = 'invalid'
+    assert_equal(get_all_errors(content, submission), {'SQ1-1g': 'invalid_format'})
+
+
+def test_character_limit_errors():
+    cases = [
+        ("SQ1-1a", 255),
+        ("SQ1-1cii", 255),
+        ("SQ1-1d", 255),
+        ("SQ1-1i-ii", 4096),
```

Comment:
We should review these character limits. They were copied from Bravo Solution™ and feel quite arbitrary—they've been removed from the content already.